### ISSUE TITLE
"Assign instructions cannot occur inside statements" without inside statements

### ISSUE THREAD

**Issue by tueda** [vermaseren/form#135]
I'm not sure whether the following error (or message) makes sense.

```
L F =
  #do i=1,10
    #$x = `i';
    + `$x'
  #enddo
;
.end
```

```
FORM 4.1 (Oct 18 2016, v4.1-20131025-249-gc11ae7c) 64-bits  Run: Thu Oct 27 19:47:43 2016
    L F =
      #do i=1,10
        #$
test.frm Line 3 ==> Assign instructions cannot occur inside statements
test.frm Line 3 ==> Missing ; ?
    x = `i';
        + `$x'
test.frm Line 4 ==> Undefined variable $x used as preprocessor variable
Program terminating at test.frm Line 4 --> 
```

**Comment by vermaseren** [comment]
Hi Takahiro,

It makes sense, once you understand what happens.
To compile the #$ statement the preprocessor needs the compiler.
But the compiler is ‘busy’ as it is in the middle of another statement.
Hence the first error message. Then it tries to guess that you
may not have finished your statement, hence the second error.
And because your #$ statement never got compiled the $ was not properly ‘cataloged’.

Technically it is not yet compiling the L F= statement but it is collecting it in the 
input buffer that will be sent to the compiler and there is only one buffer for this.
One could try to make more than one buffer (maybe moderately messy) but then
of course the next question would be “dollars inside dollars?”
There is also another problem:
If your #$ statement takes more than one line, how does that go?
All the way till the first ; ?
It requires some very careful thinking to make the syntax reasonably OK and flexible.
Maybe indeed good to think about this.

Cheers

Jos

**Comment by tueda** [comment]
@vermaseren OK. I see the problem and it will be messy to add another buffer if we are using C.

In practice, the problem here is that one can rewrite the example as follows using multiple $-variables:
```
#do i=1,10
  #$x`i' = `i';
#enddo
L F =
#do i=1,10
  + `$x`i''
#enddo
;
.end
```
but then it means that the memory consumption increases when the upper bound of the loop increases. (Suppose that it is not a constant but increases as the complexity of the problem does).

**Comment by vermaseren** [comment]
Hi Takahiro,

It is possible to make layers of buffers here. It is just that it needs very careful programming after which
it should give the functionality you were hoping for and possibly more.
It is one more incidence of that when the $’s were invented, it was not yet possible to see in which
direction it would all develop. All that was known was that it seemed useful and I could already
think of some applications. Now we have much more sophisticated programs and we know what
limits us (now). Hence we can make much more. But it takes time. Or more people.

Cheers

Jos

**Comment by tueda** [comment]
Resolved by ad8f4b7a0d and 896685e244.
